crates/sui-framework/packages/sui-framework/sources/crypto/poseidon.move

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Module which defines instances of the poseidon hash functions.
+module sui::poseidon {
+
+    #[allow(unused_const)]
+    /// Error if any of the inputs are larger than or equal to the BN254 field size.
+    const ENonCanonicalInput: u64 = 0;
+
+    #[allow(unused_const)]
+    /// Error if an empty vector is passed as input.
+    const EEmptyInput: u64 = 1;
+
+    /// @param data: Vector of BN254 field elements to hash.
+    ///
+    /// Hash the inputs using poseidon_bn254 and returns a BN254 field element.
+    ///
+    /// Each element has to be a BN254 field element in canonical representation so it must be smaller than the BN254
+    /// scalar field size which is 21888242871839275222246405745257275088548364400416034343698204186575808495617.
+    native public fun poseidon_bn254(data: &vector<u256>): u256;
```

Comment:
We have all sorts of places where we cannot do that, particularly with dynamic fields and dynamic type checks. 

So it would be nice as a general principle? Though not strictly necessary. Quite amusingly for dynamic fields we had to merge native functions for the prover (if I recall correctly). Or at least do even less logic in Move than we originally had. 

That being said... It feels like it should be fairly easy to split this into two separate native functions if we want?

To me, we could do the to_le_bytes conversion as one native function, and the hash as another. The rest of the vector manipulation would definitely be easier to grok in Move. I have a feeling though we probably haven't done this in the rest of the crypto functions? So I'm not sure if we should be consistent or not

Reply:
We haven't done this in the other crypto functions, no, but I think it makes sense to do here anyway. I'll update the PR.

Reply:
I've refactored such that all checks are done in move. Let me know what you think @sblackshear @tnowacki